### ISSUE TITLE
Contributing to PouchDB - First timer perspective

### ISSUE THREAD

**Issue by dharders** [pouchdb/pouchdb#5346]
Hey guys, first time contributor wannabe here!

I've filed a few issues on PouchDB before and contributed in some other discussions, but have never submitted an actual Pull Request. It's not that I don't know how to submit a PR, but that I am running into early build / test problems following your [Contributing] (https://github.com/pouchdb/pouchdb/blob/master/CONTRIBUTING.md) & [Testing] (https://github.com/pouchdb/pouchdb/blob/master/TESTING.md) guides.

So instead of losing time (I've lost days!) trying to get a baseline PouchDB dev branch to install, build and pass all tests BEFORE I have made a single line of change I just give up and create a new issue with Gist or Repo of my solution. This is of course inefficient and creates a greater burden on you guys.

Assume I have just performed

```
git clone https://github.com/pouchdb/pouchdb.git
cd pouchdb
npm install
```

First issue: The script `npm run build-modules` that is automatically run with `npm install` fails with 

```
...
  V wrote pouchdb-utils/lib/index.js in node mode
  V wrote pouchdb-utils/lib/index-browser.js in browser mode
build error
TypeError: Cannot read property '0' of null
    at F:\source\pouchdb\bin\build-pouchdb.js:100:46
    at Array.<anonymous> (F:\source\pouchdb\node_modules\lie\lib\index.js:88:21)
    at nextTick (F:\source\pouchdb\node_modules\immediate\lib\index.js:61:18)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)

npm ERR! Windows_NT 6.1.7601
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\User\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "run" "build-modules"
npm ERR! node v5.10.1
npm ERR! npm  v3.8.6
npm ERR! code ELIFECYCLE
npm ERR! pouchdb-monorepo@0.0.0 build-modules: `node bin/build-modules.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the pouchdb-monorepo@0.0.0 build-modules script 'node bin/build-modules.js'.
```

This issue has only popped up recently (it didn't always) and may be due to all the changes with monorepo / lerna / Leveldown.

Next Issue: `npm test` fails.

Assume I work out the above issue and the install finishes correctly (has happened with earlier versions) and I have a local CouchDB 1.6.1 server listening on port 5984 in admin party mode with CORS enabled. Running `npm test` fails.

**So my biggest question is**: 

At all times, should the **master** branch of PouchDB pass all tests (assuming I have the correct test environment setup) or is there a chance that merges are made to **master** that don't pass all tests ? 

Note: I am using a Windows 7 x64 dev environment. But I have also experienced the same issues in a linux VM (Debian wheezy).

**Comment by dharders** [comment]
OK to take Microsoft out of the equation (some Windows love would very much be appreciated Nolan!), I fired up my linux VM again.

Points of Interest:

- I solved the `npm install` issues - It turns out I had npm cache poisoning. `sudo npm cache clear` didn't solve it but deleting my `~/.npm` folder and then running `npm install` again worked ! I'm yet to try this on Windows...
- `npm run build` worked
- `npm test` first failed at:

```
1) test.all_docs.js-http Testing updating in changes:
     Uncaught TypeError: Cannot set property 'updated' of undefined
      at tests/integration/test.all_docs.js:192:25
      at packages/pouchdb/lib/index.js:127:11
      at _combinedTickCallback (internal/process/next_tick.js:67:7)
      at process._tickCallback (internal/process/next_tick.js:98:9)
```
- I got passed this by running a local `npm install leveldown` as suggested here in #5257 .
- `npm test` then ran for a bit then failed at:
```
1) suite2 test.attachments.js-http #2771 allDocs() 4, mix of atts and no atts:
     TypeError: Cannot read property 'foo.png' of undefined
      at tests/integration/test.attachments.js:1708:36
      at Array.map (native)
      at tests/integration/test.attachments.js:1705:36
      at process._tickCallback (internal/process/next_tick.js:103:7)
```
- I then ran `npm test` again and it failed early at
```
1) test.all_docs.js-http test "key" option:
     Uncaught AssertionError: key option returned 1 doc: expected [] to have a length of 1 but got 0
      at tests/integration/test.all_docs.js:529:32
      at packages/pouchdb/lib/index.js:127:11
      at _combinedTickCallback (internal/process/next_tick.js:67:7)
      at process._tickCallback (internal/process/next_tick.js:98:9)
```
- The above failure was very random (it passed just before)...then I realised my linux VM was using CouchDB 2.0dev in a 3 node cluster...
- I then pointed my tests against CouchDB server running on my windows box (1.6.1) using `COUCH_HOST=http://192.168.1.180:5984 npm test` and everything magically PASSED!

So at least now I have a baseline method for the future!

**Last Points:**
- CouchDB 2.0dev in 3 node cluster gave me grief
- CouchDB 2.0dev 1 node - add-cors-to-couchdb gave me an error and didn't work
- ` SERVER=pouchdb-server npm test` required me to **npm install sqlite3** first, like leveldown
- testing against pouchdb-server still failed after:
```
677 passing (54s)
  2 pending
  1 failing

  1)  "after each" hook for "Compact two times more complicated tree":
     Error: ETIMEDOUT
      at onError (packages/pouchdb/lib/index.js:5154:18)
      at Request._callback (packages/pouchdb/lib/index.js:5196:14)
      at self.callback (packages/pouchdb/node_modules/request/request.js:200:22)
      at Timeout._onTimeout (packages/pouchdb/node_modules/request/request.js:775:12)
```

So at least now I know (for me at least):

- Always build with linux
- Only test against CouchDB 1.6.1

Now I don't have an excuse to submit a PR ;)

**Comment by dharders** [comment]
Follow up:

- Cleaned Windows npm cache. No change...still errors out 
- Windows unable to `npm install` without failing. i.e. fails at `npm run build-modules`
- Same failure point on Win 7 and Win 10 (both x64)

```
  V wrote pouchdb-utils/lib/index-browser.js in browser mode
build error
TypeError: Cannot read property '0' of null
    at F:\source\pouchdb\bin\build-pouchdb.js:100:46
    at Array.<anonymous> (F:\source\pouchdb\node_modules\lie\lib\index.js:88:21)
    at nextTick (F:\source\pouchdb\node_modules\immediate\lib\index.js:61:18)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```
Please help @nolanlawson

**Comment by dharders** [comment]
Solved the above `npm install` error for windows

As mentioned in the error log: line 100 of build-pouchdb.js
```
filename.match(/packages\/pouchdb\/.*/)[0]);
```
is searching for forward arrows (ala Unix) and Windows uses back arrows.

```
filename.replace(/\\/g, '/')  // convert all \ arrows to /
```
This did the trick

Running `npm test` fails as bash is not Win 7 friendly (Win 10 soon with anniversary update hopefully)

Running `npm test` in Git Bash (windows bash client installed with git) got pretty far until:

```
 1) migration one should work:
     Uncaught OpenError: IO error: F:\source\pouchdb\tmp\_pouch_oldStyle\documen: The filename, directory name, or volume label syntax is incorrect.

      at F:\source\pouchdb\packages\pouchdb\node_modules\levelup\lib\levelup.js:119:34
      at F:\source\pouchdb\node_modules\leveldown\node_modules\abstract-leveldown\abstract-leveldown.js:39:16
```

**Comment by dharders** [comment]
OK, for now I'll just build with Linux. I'll revisit this again in the future.

**Comment by nolanlawson** [comment]
Yeah, sorry for the lack of movement on this. I now work for Microsoft, but I'm still getting used to the Windows ecosystem and trying to figure things out. TBH I'm still trying to decide if I want to use Cmder, Git Bash, Cygwin, Linux Subsystem, etc. But in general, the PouchDB project just doesn't work on Windows; historically nearly every single dev has been Mac or Linux.

One issue may be long paths. It looks like you're using npm 2 instead of npm 3, which has super-long paths that cause bugs in Windows. Node 5 and npm 3 are recommended.

Again, sorry for the headache. One of the dirty secrets of webdev is that it's basically a Mac monoculture, and when you use Windows you're taking a burden upon yourself, leading to things like 3 days of lost time. I totally feel your pain and will continue to feel it as I get used to Windows myself, though, so keep fighting the good fight. :)

Also FWIW for everyday tech issues a good resource is the PouchDB Slack channel or IRC channel – we usually respond faster there than on GitHub.

**Comment by dharders** [comment]
Thanks @nolanlawson ! I'll try the Slack channel next time I'm stuck.

P.S. Win 10 bash (Ubuntu on Windows) is looking very promising. It's not quite there yet, but very close! (you need to be part of the Insider program to preview it). Once it lands, it should solve all Windows build headaches for us!
